src/runtime/hexagon/device/hexagon_device_target.cc

```diff
@@ -0,0 +1,456 @@
+  if (rc != AEE_SUCCESS) {
+    TVM_LOGE_HT("failed to open channel rc=0x%x", rc);
+  } else {
+    count_channel_open_++;
+    TVM_LOGD_HT("channel open success and rpcmem_init done");
+  }
+  return rc;
+}
+
+int HexagonTarget::CloseDomainChannel() {
+  if (domain_channel_handle_ == AEE_EUNKNOWN) return AEE_SUCCESS;
+
+  const DspRpcAPI* dsp_api = DspRpcAPI::Global();
+  const StubAPI* stub_api = StubAPI::Global();
+
+  int rc = stub_api->tvm_hexagon_remote_close(domain_channel_handle_);
+  if (rc == AEE_SUCCESS) {
+    domain_channel_handle_ = AEE_EUNKNOWN;
+    if (auto* deinit_p = dsp_api->rpcmem_deinit_ptr(true)) deinit_p();
+    TVM_LOGD_HT("channel close success and rpcmem_deinit done");
+  } else {
+    TVM_LOGE_HT("failed to close domain channel rc=0x%x", rc);
+  }
+  return rc;
+}
+
+void HexagonTarget::ReleaseLibrary() {
+  crit_section_.lock();
+  if (module_pointer_ != AEE_EUNKNOWN) {
+    const StubAPI* stub_api = StubAPI::Global();
+    int rc = stub_api->tvm_hexagon_remote_release_library(
+        domain_channel_handle_, module_pointer_);
+    if (rc != AEE_SUCCESS) {
+      TVM_LOGE_HT("failed to unload device library rc=0x%x", rc);
+    } else {
+      module_pointer_ = AEE_EUNKNOWN;
+    }
+  }
+  crit_section_.unlock();
+}
+
+void* HexagonTarget::Alloc(unsigned size, unsigned align) {
+  const DspRpcAPI* dsp_api = DspRpcAPI::Global();
+
+  // Opening the domain channel should be done once.
+  crit_section_.lock();
+  int rc_oc = OpenDomainChannel(/*use_unsigned_pd*/ true);
+  crit_section_.unlock();
+  if (rc_oc != AEE_SUCCESS) {
+    TVM_LOGE_HT("mem alloc failed: unable to open domain channel");
+    return nullptr;
+  }
+
+  void* mem =
+      dsp_api->rpcmem_alloc_ptr()(RPCMEM_HEAP, RPCMEM_DEFAULT_FLAGS, size);
+  if (mem == nullptr) {
+    TVM_LOGE_HT("mem alloc failed for size=0x%x alignment=0x%x", size, align);
+    return nullptr;
+  }
+  int mem_fd = dsp_api->rpcmem_to_fd_ptr()(mem);
+  uintptr_t dsp_va = 0;
+  int rc = dsp_api->remote_mmap64_ptr()(
+      mem_fd, 0, reinterpret_cast<uintptr_t>(mem), size, &dsp_va);
+  if (rc != AEE_SUCCESS) {
+    TVM_LOGE_HT("buffer mapping failed for remote_map64 fd=0x%x rc=0x%x",
+                mem_fd, rc);
+    return nullptr;
+  }
+
+  void* dsp_addr = reinterpret_cast<void*>(dsp_va);
+  AddAddrMapping(dsp_addr, mem, size);
+  return dsp_addr;
+}
+
+void HexagonTarget::Free(void* ptr) {
+  const DspRpcAPI* dsp_api = DspRpcAPI::Global();
+
+  auto aa = GetAppsAddr(ptr, true);
+  if (aa.first == nullptr) return;
+
+  int rc = dsp_api->remote_munmap64_ptr()(reinterpret_cast<uintptr_t>(ptr),
+                                          aa.second);
+  if (rc != AEE_SUCCESS) {
+    TVM_LOGE_HT("buffer unmapping failed rc=0x%x", rc);
+  }
+  RemoveAddrMapping(ptr);
+  dsp_api->rpcmem_free_ptr()(aa.first);
+}
+
+void HexagonTarget::CopyDeviceToDevice(void* dst, const void* src,
+                                       unsigned len) {}
```

Comment:
When we execute model (such as NNVM frontend / Relay frontend), we will meet CopyDeviceToDevice, imagine we have `module.get_output(0, tvm.nd.empty(output_shape, dtype="float32", ctx=ctx)).asnumpy()`. The implementation is not difficult (Using `GetAppsAddr` to get `src / dst app addr` and `memcpy`). Leave this comment reminding us we should implement it and don't leave it empty. Otherwise, we will get 0 of output.